src/main/scala/nl.knaw.dans.easy.deposit/servlets/DepositServlet.scala

```diff
@@ -157,11 +159,49 @@ class DepositServlet(app: EasyDepositApiApp)
       ).getOrRecoverResponse(respond)
   }
 
-  private def postFileItem(uuid: UUID, path: Path, item: FileItem): Try[(Boolean, FileItem)] = {
-    val fullPath = path.resolve(item.name)
-    managed(item.getInputStream)
-      .apply(app.writeDepositFile(_, user.id, uuid, fullPath))
-      .map((_, item))
+  private def uploadFileItem(uuid: UUID, path: Path, uploadItem: FileItem): Try[(Boolean, FileItem)] = {
+    val extensionIsZip = uploadItem.name.matches(".*.g?z(ip)?")
+    lazy val contentTypeIsZip = uploadItem.contentType.exists(_.matches(
+      "(application|multipart)/(x-)?g?zip(-compress(ed)?)?( .*)?"
+    ))
+    logger.debug(s"is it a ZIP? ${ uploadItem.name } : $extensionIsZip; ${ uploadItem.contentType } : $contentTypeIsZip ")
+    if (extensionIsZip || contentTypeIsZip)
+      managed(uploadItem.getInputStream)
+        .apply(uploadZip(uuid, path, uploadItem, _))
+        .map((_, uploadItem))
+    else {
+      val fullPath = path.resolve(uploadItem.name)
+      managed(uploadItem.getInputStream)
+        .apply(app.writeDepositFile(_, user.id, uuid, fullPath))
+        .map((_, uploadItem))
+    }
+  }
+
+  private def uploadZip(uuid: UUID, path: Path, uploadItem: FileItem, uploadInputStream: InputStream): Try[Boolean] = {
+    val zipInputStream = uploadItem.charset
+      .map(charSet => new ZipInputStream(uploadInputStream, Charset.forName(charSet)))
+      .getOrElse(new ZipInputStream(uploadInputStream))
+
+    def handleZipEntry(zipEntry: ZipEntry) = {
+      if (zipEntry.isDirectory) Success(false)
+      else {
+        logger.info(s"extracting ${ zipEntry.getName } size=${ zipEntry.getSize } crc=${ zipEntry.getCrc } method=${ zipEntry.getMethod }")
+        val fullPath = path.resolve(zipEntry.getName)
+        app.writeDepositFile(zipInputStream, user.id, uuid, fullPath)
+      }
+    }
+
+    while (
+      Try(zipInputStream.getNextEntry)
+        .map(Option(_).map(handleZipEntry)) match {
+        case Success(None) => false // end of zip
+        case Success(Some(Success(_))) => true // extracted and uploaded
+        case Success(Some(Failure(e))) => return Failure(e) // could not save
+        case Failure(e) if e.isInstanceOf[ZipException] => // could not extract TODO still fail fast? Other files might hav been uploaded.
+          return Failure(BadRequestException(s"ZIP file is malformed. ${ uploadItem.name } $e"))
+        case Failure(e) => return Failure(e)
+      }) {}
+    Success(false)
```

Comment:
do you always return `false`??? Why is that? It doesn't make much sense to me. If you just want to communicate `Success` or `Failure`, why not use `Try[Unit]` here?